Implement bot protection login in target branches

pkong-ds opened the pull request:
## Blocked by
- #4395 
## What's in this PR?
Implement bot protection in below branches
- [x] signup flow
  - [x] identify
    - [x] loginid
    - [x] oauth
  - [ ] create_authenticator
    - [ ] password
    - [ ] oobotp
    - [ ] totp
- [x] promote flow
  - [x] identify
    - [x] loginid
    - [x] oauth
  - [ ] create_authenticator (handled in signupflow > create_authenticator)
- [x] login flow
  - [x] identify
    - [x] loginid
    - [x] oauth
    - [x] passkey
  - [x] authenticate
    - [x] password
    - [x] passkey
    - [x] oobotp
    - [x] totp
    - [x] recoverycode
- [x] signup_login flow
  - [x] switch flow success
  - [x] identify
    - [x] loginid
    - [x] oauth
    - [x] passkey
- [x] reauth flow
  - [x] authentication
    - [x] password
    - [x] passkey
    - [x] oobotp
    - [x] totp
- [x] account_recovery_flow
  - [x] identify

pkong-ds commented:
@louischan-oursky Can I have your attention to these 2 commits? Gonna replicate this implementation across all other steps and branches, please advise if any!

```md
Bot protection is supported only in the following step types:

    identify in signup, promote, login, signup_login, and account_recovery.
    create_authenticator in signup and promote.
    authenticate in login and reauth.


```

[Implement bot protection input in input_step_identify](https://github.com/authgear/authgear-server/pull/4415/commits/d5cd1af69b56e390ceae193b060209c670008a21)
[Implement bot protection in login identify by loginid](https://github.com/authgear/authgear-server/pull/4415/commits/15c4c8dcb5106209cffddc172e18b770956dfa04)

<img width="920" alt="image" src="https://github.com/authgear/authgear-server/assets/74223769/642c6369-6279-4db8-8514-0d1ff6bca662">

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -138,7 +138,19 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
-			err = nil
+			uw, isJoinedError := err.(interface{ Unwrap() []error })
+			if isJoinedError {
+				errs := uw.Unwrap()
```
Use errors.Is or errors.As

pkong-ds replied:
As per offline discussion, will use newly added `errorutil.Partition`

ref 

```go
func Partition(err error, predicate func(err error) bool) (matched error, notMatched error) {
	var _matched []error
	var _notMatched []error
	Unwrap(err, func(err error) {
		b := predicate(err)
		if b {
			_matched = append(_matched, err)
		} else {
			_notMatched = append(_notMatched, err)
		}
	})
	return errors.Join(_matched...), errors.Join(_notMatched...)
}
```

louischan-oursky replied:
One thing you need to pay attention. After partition, the order of the error should stay the same.

pkong-ds commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -108,6 +113,37 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
+			// We still consider the flow has something changes.
+			changed = true
+
+			return &AcceptResult{
```
As per offline discussion,
Early return is fine for below
- VerifyFailed
- VerifyServiceUnavailable
While
- VerifySuccess
should keep processing the input

pkong-ds commented on pkg/lib/authenticationflow/declarative/input_take_bot_protection.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// This input should be dynamically added, but not used directly.
+type InputTakeBotProtection struct {
```
> The input should be under the key bot_protection.

IMO no

The expected usage is below

```go
// b is an existing schema builder without `bot_protection` field
b = AddBotProtectionToExistingSchemaBuilder(b)
```

where

```go
func AddBotProtectionToExistingSchemaBuilder(sb validation.SchemaBuilder) validation.SchemaBuilder {
	sb.AddRequired("bot_protection")
	sb.Properties().Property(("bot_protection"), NewInputTakeBotProtectionSchemaBuilder())
	return sb
}
```

pkong-ds commented on pkg/lib/authenticationflow/declarative/input_step_identify.go:
```diff
@@ -81,6 +85,13 @@ func (i *InputSchemaStepIdentify) SchemaBuilder() validation.SchemaBuilder {
+		if option.isBotProtectionRequired() {
+			// bot_protection is required.
+			required = append(required, "bot_protection")
+			b.Properties().Property("bot_protection", NewInputTakeBotProtectionSchemaBuilder())
```
I recalled why I made the decision of NOT using `AddBotProtectionToExistingSchemaBuilder` 🙏 

Observing `InputSchemaStepIdentify.SchemaBuilder()`, we can see that it utilized the helper function `setRequiredAndAppendOneOf` to update the output schema `required` field.

If I use `AddBotProtectionToExistingSchemaBuilder`, which use `AddRequired` under the hood, the added required field `bot_protection` might get overriden by other calls of `setRequiredAndAppendOneOf`.

Hence, I decided to follow the pattern here to write another helper function `setRequired`. Please advise if otherwis

pkong-ds commented:
@louischan-oursky all logic should be implemented 🙏 now working on e2e tests (and manual testing for passkey, etc.)

Feel free to review first, changes since last review should start from 855e7eba88f7bb62f8882860c7f65d12ddeb7888

louischan-oursky commented on pkg/lib/botprotection/cloudflare_client.go:
```diff
@@ -0,0 +1,84 @@
+		formValues.Add("remoteip", remoteip)
+	}
+
+	resp, err := c.HTTPClient.PostForm(CloudflareVerifyEndpoint, formValues)
```
You need to use defer to close the response body. The pattern is 

```
resp, err := c.HTTPClient.SomeMethodReturningResponse()
if err != nil {
  ...
}
defer resp.Body.Close()
```

louischan-oursky commented on pkg/lib/botprotection/cloudflare_client.go:
```diff
@@ -0,0 +1,84 @@
+		err := errors.Join(
+			fmt.Errorf("unrecognised response body from cloudflare turnstile"),
+			err,
+			ErrVerificationFailed,
```
Separate the error handling.

```golang
httpBodyBytes, err := io.ReadAll(resp.Body)
if err != nil {
  return nil, errors.Join(ErrServiceUnavailable, fmt.Errorf("failed to read response body: %w", err))
}
err = json.Unmarshal(httpBodyBytes, &respBody)
if err != nil {
  // Return a fmt.Errorf error to means this is a internal server error that cannot be handled.
  return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
}
if respBody.Success == nil {
  // Also a fmt.Errorf error to mean this is an internal server error that cannot be handled.
  return nil, fmt.Errorf("unexpected response body: %v", string(httpBodyBytes))
}
```

louischan-oursky commented on pkg/lib/botprotection/cloudflare_client.go:
```diff
@@ -0,0 +1,84 @@
+	failedResp := &CloudflareTurnstileErrorResponse{
+		ErrorCodes: respBody.ErrorCodes,
+	}
+	err = errors.New(failedResp.Error())
```
Do not do this. errors.New returns an opaque error. failedResp is already an error that you can return directly.

louischan-oursky commented on pkg/lib/botprotection/cloudflare_client.go:
```diff
@@ -0,0 +1,84 @@
+	}
+	err = errors.New(failedResp.Error())
+	for _, errCode := range CloudFlareTurnstileServiceUnavailableErrorCodes {
+		if strings.Contains(failedResp.Error(), string(errCode)) {
```
You can just loop ErrorCodes. No need to call failedResp.Error(). In general, you rarely can .Error() directly.

louischan-oursky commented on pkg/lib/botprotection/cloudflare_client.go:
```diff
@@ -0,0 +1,84 @@
+		}
+	}
+
+	return nil, errors.Join(err, ErrVerificationFailed)
```
This is an internal server error. We receive a response we do not know how to interpret. 

```
return nil, fmt.Errorf("unexpected response body: %v", string(httpBodyBytes))
```

louischan-oursky commented on pkg/lib/botprotection/cloudflare_response.go:
```diff
@@ -0,0 +1,62 @@
+	ErrorCodes []CloudflareTurnstileErrorCode `json:"error-codes"`
+
+	// specific to Success == true
+	ChallengeTs string `json:"challenge_ts"` // ISO timestamp for the time the challenge was solved.
```
Add omitempty to ALL fields.

louischan-oursky commented on pkg/lib/botprotection/cloudflare_response.go:
```diff
@@ -0,0 +1,62 @@
+// parsed error response
+type CloudflareTurnstileErrorResponse struct {
+	ErrorCodes []CloudflareTurnstileErrorCode `json:"error-codes"`
+}
```
I think about this again. We can reuse CloudflareTurnstileRawResponse to represent both response. When it is successfully, we return `(*CloudflareTurnstileRawResponse, nil)`, When it is an error, we return `(nil, error)` when the error COULD be `*CloudflareTurnstileRawResponse`

louischan-oursky commented on pkg/lib/botprotection/provider.go:
```diff
@@ -44,17 +45,14 @@ func (p *Provider) verifyTokenByCloudflare(token string) error {
-		return ErrVerificationFailed
+	if successResp == nil {
+		err = fmt.Errorf("cloudflare no error but no success response")
+		return errors.Join(err, ErrVerificationFailed)
```
No need this checking. If err is nil, then resp must be present. It is a programming error if this does not hold. Programming error is supposed to be caught by unit tests. We do not handle programming error this way. And of course, we do not report programming error as ErrVerificationFailed. They are entirely different.

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -114,47 +114,50 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
+						}
 					}
+				} else {
+					err = nil
```
Does not seem correct. Why `else err =nil`?

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -114,47 +114,50 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
+						Outcome: BotProtectionVerificationOutcomeVerified,
+					}}
+			case ErrorBotProtectionVerificationStatusFailed:
+				err = nil
```
Add a comment to explain why we want to set `err = nil` to continue the accept loop.

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -114,47 +114,50 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
+						Outcome: BotProtectionVerificationOutcomeFailed,
+					}}, botprotection.ErrVerificationFailed
+			case ErrorBotProtectionVerificationStatusServiceUnavailable:
+				err = nil
```
Add a comment to explain why we want to set `err = nil` to continue the accept loop.

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -114,47 +114,50 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
+					}}, botprotection.ErrVerificationServiceUnavailable
+			default:
+				// do nothing if unrecognized status
 				err = nil
```
You panic when you see something unexpected. Do not ignore the situation.

louischan-oursky commented on pkg/util/errorutil/partition_test.go:
```diff
@@ -0,0 +1,68 @@
+		errA := errors.New("a")
+		matched, notMatched := errorutil.Partition(errA, func(err error) bool { return errors.Is(err, errA) })
+		So(errors.Is(matched, errA), ShouldBeTrue)
+		So(errors.Is(notMatched, errA), ShouldBeFalse)
```
Add one assertion to assert `notMatched` is nil.

louischan-oursky commented on pkg/util/errorutil/partition_test.go:
```diff
@@ -0,0 +1,68 @@
+		errNotA := errors.New("not a")
+		matched, notMatched := errorutil.Partition(errA, func(err error) bool { return errors.Is(err, errNotA) })
+		So(errors.Is(matched, errA), ShouldBeFalse)
+		So(errors.Is(notMatched, errA), ShouldBeTrue)
```
Add one assertion to assert `matched` is nil.

louischan-oursky commented on pkg/util/errorutil/partition_test.go:
```diff
@@ -0,0 +1,68 @@
+		So(errors.Is(notMatched, errD), ShouldBeTrue)
+		So(errors.Is(notMatched, errE), ShouldBeTrue)
+		So(errors.Is(notMatched, errF), ShouldBeTrue)
+	})
```
Add one more test case to test the order of the partitioned errors.

```
errA := errors.New("a")
errB := errors.New("b")
errC := errors.New("c")
errD := errors.New("d")
errE := errors.New("e")
errF := errors.New("f")

joined := errors.Join(errA, errB, errC, errD, errE, errF)

matched, notMatched := errorutil.Partition(joined, func(err error) bool {
	return errors.Is(err, errC)
})

var unwrapOrderHistory strings.Builder

Unwrap(notMatched, func(err error) {
	unwrapOrderHistory.WriteString(err.Error())
})
So(unwrapOrderHistory.String(), ShouldEqual, "abdef")

```

pkong-ds replied:
Already added in later commits.

louischan-oursky commented on pkg/lib/authenticationflow/accept.go:
```diff
@@ -119,20 +119,11 @@ func accept(ctx context.Context, deps *Dependencies, flows Flows, inputFn func(i
-				}
+					return errors.As(err, &_errBPV) && _errBPV.Status == ErrorBotProtectionVerificationStatusSuccess
+				})
+				err = notMatched
```
line 122 to 126 should be applied to all 3 cases, not just the the success case. The rationale is that we want to extract the special error from the error chain, and set err to the rest of the error chain.

louischan-oursky commented on pkg/lib/authenticationflow/declarative/intent_create_authenticator_password.go:
```diff
@@ -81,6 +81,9 @@ func (i *IntentCreateAuthenticatorPassword) ReactTo(ctx context.Context, deps *a
 			}
+			if !IsBotProtectionSpecialErrorSuccess(bpSpecialErr) {
+				return nil, bpSpecialErr
+			}
```
Instead of adding IsBotProtectionSpecialErrorSuccess, maybe we should change how HandleBotProtection behave.

1. It returns `(nil, err)`. So that we will `if err != nil { return nil, err }` early. Note that err COULD be specialErr.
2. It returns `(specialErr, nil)`. So that the special error is returned along with the normal return value.

What do you think?

pkong-ds replied:
I think this ~10-12 lines of re-appearing code should be refactored as a single util function, which has your mentioned return interface 🙏

pkong-ds replied:
On second thought, I just moved `IsBotProtectionSpecialErrorSuccess` checking into `HandleBotProtection` as you suggested.

### Rationale
IMO Checking `isBPRequired` and actually performing `verifyBP` should be 2 separate function calls.

louischan-oursky commented on e2e/tests/bot_protection/README.md:
```diff
@@ -0,0 +1,63 @@
+> Bot protection is supported only in the following step types:
+
+    identify in signup, promote, login, signup_login, and account_recovery.
+    create_authenticator in signup and promote.
```
Remove this line.

louischan-oursky commented on e2e/tests/bot_protection/signup/identify/loginid/error_missing_input.test.yaml:
```diff
@@ -0,0 +1,53 @@
+name: Signup with Bot Protection - Identify - Missing Bot Protection Input
```
Instead of nested the directories, we can just use a more flat structure `signup_identify_loginid_error_missing_input.test.yaml`

louischan-oursky replied:
Ignore this comment

pkong-ds replied:
(Because fixtures would be flattened to the same folder)

louischan-oursky commented on pkg/lib/config/bot_protection_env.go:
```diff
@@ -0,0 +1,14 @@
+package config
+
+type BotProtectionEnvironmentConfig struct {
+	CloudflareEndpoint  string `envconfig:"BOT_PROTECTION_CLOUDFLARE_ENDPOINT" default:"https://challenges.cloudflare.com/turnstile/v0/siteverify"`
```
First of all, do not put the real value here. Keep the real value in the original place. If the envvar is set, then use the envvar, otherwise, keep using the real value.

louischan-oursky commented on pkg/lib/config/bot_protection_env.go:
```diff
@@ -0,0 +1,14 @@
+}
+
+// NewDefaultBotProtectionEnvironmentConfig provides default bot protection config
+func NewDefaultBotProtectionEnvironmentConfig() *BotProtectionEnvironmentConfig {
```
Remove this function. We define the real endpoint in a SINGLE place.

louischan-oursky commented on pkg/lib/config/environment.go:
```diff
@@ -71,6 +71,8 @@ type EnvironmentConfig struct {
 	// RedisConfig configures the redis connection config
 	RedisConfig RedisEnvironmentConfig `envconfig:"REDIS_CONFIG"`
 
+	BotProtectionConfig BotProtectionEnvironmentConfig `envconfig:"BOT_PROTECTION"`
```
Rename this to `End2EndBotProtection`, `End2EndBotProtectionConfig`, `E2E_BOT_PROTECTION`

louischan-oursky commented:
@pkong-ds How WIP is this PR?

pkong-ds commented:
@louischan-oursky ready for review 🙏 local CI & e2e pass

Note changes since last review should start from eb93d249eb4896aa36768a04068b15cf55effbe3

Commit message: [Revert all create_authenticator-related bot protection changes](https://github.com/authgear/authgear-server/commit/eb93d249eb4896aa36768a04068b15cf55effbe3)

## Changes
- remove all `create_authenticator` related bot protection
- adopt misc suggestions from you
- added general e2e tests like `no need require captcha again`

pkong-ds commented on pkg/lib/authenticationflow/declarative/intent_account_recovery_flow_step_identify.go:
```diff
@@ -96,9 +96,10 @@ func (i *IntentAccountRecoveryFlowStepIdentify) CanReactTo(ctx context.Context,
+			FlowRootObject:   flowRootObject,
+			JSONPointer:      i.JSONPointer,
+			Options:          i.Options,
+			BotProtectionCfg: deps.Config.BotProtection,
```
This is unexpected 🙏 thanks for spotting

pkong-ds commented:
@louischan-oursky updated, thanks!

Changes should start from [dc0dd5e](https://github.com/authgear/authgear-server/pull/4415/commits/dc0dd5ef8a40bb04f24ef48f16a4828c2196dfc7)

Commit msg: [Fix bot protection clients should throw internal server error on unexpected response body](https://github.com/authgear/authgear-server/pull/4415/commits/dc0dd5ef8a40bb04f24ef48f16a4828c2196dfc7)

pkong-ds commented on pkg/lib/botprotection/recaptchav2_client.go:
```diff
@@ -0,0 +1,82 @@
+
+	// failed
+	if len(respBody.ErrorCodes) == 0 {
+		return nil, ErrVerificationFailed // fail without error codes if empty
```
I gave a thought on it, I still think we should return `ErrVerificationFailed` here.

> This is an unexpected response body: success is false but there is no error codes. It is definitely unexpected.

Yes, this is unexpected. But IMO a server should handle such "unexpected response" gracefully by telling the caller `Your verification somehow failed. We are not sure why coz there's no error-codes. But your verification, for sure, failed` for below response from cloudflare/recaptchav2/other captcha providers.

```json5
{
  "success": "false",
  "error-codes": []
}
```

Any thoughts?

pkong-ds replied:
I still changed according to your suggestion to prevent another round trip. 🙏  But I stand by my original stance

louischan-oursky commented on pkg/lib/authenticationflow/declarative/intent_lookup_identity_login_id.go:
```diff
@@ -96,6 +86,10 @@ func (n *IntentLookupIdentityLoginID) ReactTo(ctx context.Context, deps *authflo
+		var botProtection *InputTakeBotProtectionBody
+		if inputBP, ok := inputTakeLoginID.(inputTakeBotProtection); ok {
+			botProtection = inputBP.GetBotProtectionProvider()
+		}
```
According to offline discussion, this pattern reveals a problem.

The original intention here is to retain the input and verify again. But that is not possible. Most bot protection provider does not verify the same token twice. We have to store the verification result (which we have already done).

Remove this snippet and add a test for the following case: in signup_login_flow, if bot protection is required, then after passing bot protection and switched flow successfully, the switched flow does not require bot protection anymore.

pkong-ds commented:
@louischan-oursky updated, thanks!

louischan-oursky commented:
Thank you. This PR was tremendous.
